# --- Day 23: Coprocessor Conflagration ---

You decide to head directly to the CPU and fix the printer from there. As you get close, you find an experimental coprocessor doing so much work that the local programs are afraid it will halt and catch fire. This would cause serious issues for the rest of the computer, so you head in and see what you can do.

The code it's running seems to be a variant of the kind you saw recently on that tablet. The general functionality seems very similar, but some of the instructions are different:

- set X Y sets register X to the value of Y.
- sub X Y decreases register X by the value of Y.
- mul X Y sets register X to the result of multiplying the value contained in register X by the value of Y.
- jnz X Y jumps with an offset of the value of Y, but only if the value of X is not zero. (An offset of 2 skips the next instruction, an offset of -1 jumps to the previous instruction, and so on.)

Only the instructions listed above are used. The eight registers here, named a through h, all start at 0.

The coprocessor is currently set to some kind of debug mode, which allows for testing, but prevents it from doing any meaningful work.

**If you run the program (your puzzle input), how many times is the mul instruction invoked?**

In [26]:
# the puzzle input
with open('puzzle_inputs/day23_input.txt') as f:
    data = f.read().strip().split("\n")
puzzle_input = [line for line in data]
puzzle_input[:10]

['set b 57',
 'set c b',
 'jnz a 2',
 'jnz 1 5',
 'mul b 100',
 'sub b -100000',
 'set c b',
 'sub c -17000',
 'set f 1',
 'set d 2']

In [41]:
len(puzzle_input)

32

In [31]:
from collections import defaultdict

def part_one(instructions, verbose=False):
    """goes through instructions and perfoms them as per part one"""
    
    registers = defaultdict(int)
    i = 0
    total_instructions = len(instructions)
    mul_runs = 0
    
    while i >=0 and i < total_instructions:
        
        # first, parse instruction
        f = instructions[i].split()
        if verbose: print(i, f)
        func = f[0]
        
        try:
            x = int(f[1])
        except:
            x = f[1]

        if len(f) == 3:
            try:
                y = int(f[2])
            except:
                y = f[2]
        
        # now to run the func code
        if func == "set":
            y = registers[y] if type(y) is str else y
            registers[x] = y
        elif func == "add":
            y = registers[y] if type(y) is str else y
            registers[x] += y
        elif func == "mul":
            y = registers[y] if type(y) is str else y
            registers[x] = registers[x] * y
            mul_runs += 1
        elif func == "sub":
            y = registers[y] if type(y) is str else y
            registers[x] -= y
        elif func == "jnz":
            x = registers[x] if type(x) is str else x
            if x != 0:
                i += registers[y] if type(y) is str else y
                if verbose: print(f"jumping to {i}")
                # now to skip the i += step below
                continue
                
        i += 1

    return registers, mul_runs

part_one(puzzle_input, verbose=False)

(defaultdict(int,
             {'a': 0,
              'b': 57,
              'c': 57,
              'd': 57,
              'e': 57,
              'f': 0,
              'g': 0,
              'h': 1}),
 3025)

`3025` is the answer for part one

# --- Part Two ---

Now, it's time to fix the problem.

The debug mode switch is wired directly to register a. You flip the switch, which makes register a now start at 1 when the program is executed.

Immediately, the coprocessor begins to overheat. Whoever wrote this program obviously didn't choose a very efficient implementation. You'll need to optimize the program if it has any hope of completing before Santa needs that printer working.

The coprocessor's ultimate goal is to determine the final value left in register h once the program completes. Technically, if it had that... it wouldn't even need to run the program.

After setting register a to 1, if the program were to run to completion, what value would be left in register h?

In [40]:
from collections import defaultdict

def part_one(instructions, verbose=False):
    """goes through instructions and perfoms them as per part one"""
    
    registers = defaultdict(int)
    registers["a"] = 1
    i = 0
    loops = 0
    total_instructions = len(instructions)
    
    while i >=0 and i < total_instructions:
        
        # first, parse instruction
        f = instructions[i].split()
        if verbose: print(i, f)
        func = f[0]
        
        try:
            x = int(f[1])
        except:
            x = f[1]

        if len(f) == 3:
            try:
                y = int(f[2])
            except:
                y = f[2]
        
        # now to run the func code
        if func == "set":
            y = registers[y] if type(y) is str else y
            registers[x] = y
        elif func == "add":
            y = registers[y] if type(y) is str else y
            registers[x] += y
        elif func == "mul":
            y = registers[y] if type(y) is str else y
            registers[x] = registers[x] * y
        elif func == "sub":
            y = registers[y] if type(y) is str else y
            registers[x] -= y
        elif func == "jnz":
            x = registers[x] if type(x) is str else x
            if x != 0:
                i += registers[y] if type(y) is str else y
                if verbose: print(f"jumping to {i}")
                # now to skip the i += step below
                continue
                
        i += 1
        loops += 1
        if loops % 100_000 == 0:
            print(registers)
        if loops > 10_000_000:
            break

    return registers

part_one(puzzle_input, verbose=False)

defaultdict(<class 'int'>, {'a': 1, 'b': 105700, 'c': 122700, 'f': 1, 'd': 2, 'e': 16667, 'g': 2})
defaultdict(<class 'int'>, {'a': 1, 'b': 105700, 'c': 122700, 'f': 1, 'd': 2, 'e': 33334, 'g': 33334})
defaultdict(<class 'int'>, {'a': 1, 'b': 105700, 'c': 122700, 'f': 1, 'd': 2, 'e': 50000, 'g': -5700})
defaultdict(<class 'int'>, {'a': 1, 'b': 105700, 'c': 122700, 'f': 0, 'd': 2, 'e': 66667, 'g': 66667})
defaultdict(<class 'int'>, {'a': 1, 'b': 105700, 'c': 122700, 'f': 0, 'd': 2, 'e': 83333, 'g': 60966})
defaultdict(<class 'int'>, {'a': 1, 'b': 105700, 'c': 122700, 'f': 0, 'd': 2, 'e': 100000, 'g': 2})
defaultdict(<class 'int'>, {'a': 1, 'b': 105700, 'c': 122700, 'f': 0, 'd': 3, 'e': 10968, 'g': -94732})
defaultdict(<class 'int'>, {'a': 1, 'b': 105700, 'c': 122700, 'f': 0, 'd': 3, 'e': 27635, 'g': -22798})
defaultdict(<class 'int'>, {'a': 1, 'b': 105700, 'c': 122700, 'f': 0, 'd': 3, 'e': 44301, 'g': 132903})
defaultdict(<class 'int'>, {'a': 1, 'b': 105700, 'c': 122700, 'f': 0, 'd': 3,

defaultdict(<class 'int'>, {'a': 1, 'b': 105700, 'c': 122700, 'f': 0, 'd': 14, 'e': 81612, 'g': 1036868})
defaultdict(<class 'int'>, {'a': 1, 'b': 105700, 'c': 122700, 'f': 0, 'd': 14, 'e': 98279, 'g': 14})
defaultdict(<class 'int'>, {'a': 1, 'b': 105700, 'c': 122700, 'f': 0, 'd': 15, 'e': 9247, 'g': -96453})
defaultdict(<class 'int'>, {'a': 1, 'b': 105700, 'c': 122700, 'f': 0, 'd': 15, 'e': 25914, 'g': 282995})
defaultdict(<class 'int'>, {'a': 1, 'b': 105700, 'c': 122700, 'f': 0, 'd': 15, 'e': 42580, 'g': 638700})
defaultdict(<class 'int'>, {'a': 1, 'b': 105700, 'c': 122700, 'f': 0, 'd': 15, 'e': 59247, 'g': -46453})
defaultdict(<class 'int'>, {'a': 1, 'b': 105700, 'c': 122700, 'f': 0, 'd': 15, 'e': 75914, 'g': 1032995})
defaultdict(<class 'int'>, {'a': 1, 'b': 105700, 'c': 122700, 'f': 0, 'd': 15, 'e': 92580, 'g': 1388700})
defaultdict(<class 'int'>, {'a': 1, 'b': 105700, 'c': 122700, 'f': 0, 'd': 16, 'e': 3548, 'g': 16})
defaultdict(<class 'int'>, {'a': 1, 'b': 105700, 'c': 122700, 

defaultdict(int,
            {'a': 1,
             'b': 105700,
             'c': 122700,
             'd': 17,
             'e': 81183,
             'f': 0,
             'g': 81183})